### ISSUE TITLE
duplicate entries in "Account Requests Waiting for Requester Response" table

### ISSUE THREAD

**Issue by ahelsing** [GENI-NSF/geni-ar#40]
I wasn't explicitly looking for this, but i found it and believe i can reproduce it, so i'm reporting it in case it's easy to figure out why it's happening and/or it's symptomatic of something more serious.

Steps to reproduce:
1. Request an account with a given username, say `chaos8`
2. As an admin, approve the request
3. As an admin, delete the `chaos8` account
4. Request an account with that same username, `chaos8`, again
5. As an admin, choose the "EMAIL REQUESTER" option and send e-mail to the requester
6. On [https://shib-idp1.gpolab.bbn.com/manage/display_requests.php], in the "Account Requests Waiting for Requester Response" table, chaos8 shows up twice with identical metadata.
   
   _Imported from trac ticket [#40](http://trac.gpolab.bbn.com/idp-tools/ticket/40),  created by chaos on 09-17-2013 at 15:04, last modified: 10-03-2013 at 16:53_

**Comment by ahelsing** [comment]
FYI, it looks like what's going on under the hood here is:
- When you delete an account, it winds up in state deleted in the idp_account_request table:

```
 14 | Chaos      | Test User 10 | chaos10@gpolab.bbn.com | chaos10            | 617-873-2287 | {SSHA}ELIDED | BBN Technologies | Test User |     | i want to test being a user | 2013-09-17 19:21:54.694352 |                   | 2013-09-17 19:22:29.72851 | DELETED
```
- When you later request an account with the same userid, it shows up in the table as a second entry:

```
 14 | Chaos      | Test User 10 | chaos10@gpolab.bbn.com | chaos10            | 617-873-2287 | {SSHA}ELIDED | BBN Technologies | Test User |     | i want to test being a user | 2013-09-17 19:21:54.694352 |                   | 2013-09-17 19:22:29.72851 | DELETED
 16 | Chaos      | Test User 10 | chaos10@gpolab.bbn.com | chaos10            | 617-873-2287 | {SSHA}ELIDED | BBN Technologies | Test User |     | i want to test being a user | 2013-09-17 19:28:00.337092 |                   |                           | REQUESTED
```
- When you e-mail the requester of that second account, both entries get moved to state `EMAILED_REQUESTER`:

```
 14 | Chaos      | Test User 10 | chaos10@gpolab.bbn.com | chaos10            | 617-873-2287 | {SSHA}ELIDED | BBN Technologies | Test User |     | i want to test being a user | 2013-09-17 19:21:54.694352 |                   | 2013-09-17 19:22:29.72851 | EMAILED_REQUESTER
 16 | Chaos      | Test User 10 | chaos10@gpolab.bbn.com | chaos10            | 617-873-2287 | {SSHA}ELIDED | BBN Technologies | Test User |     | i want to test being a user | 2013-09-17 19:28:00.337092 |                   |                           | EMAILED_REQUESTER
```

_Trac comment by chaos on 09-17-2013 at 15:31_

**Comment by ahelsing** [comment]
Looking at this, it makes sense that we don't want to allow duplicate requests for the same username.  So, should I:

1 - check if a request exists, in any state, for that username and reject it?  Then if an account is denied or deleted and the user wants to request it again, he/she needs to use a different username.

2 - Reject the request if a request with that username is already pending. (If its been approved, then an account will exist and it will have already been rejected for that reason.)  If the request is denied or deleted, allow it and change the state of the existing request.  This solves issue #1 but loses history.

I prefer option #1.

_Trac comment by phelinek on 09-18-2013 at 11:07_

**Comment by ahelsing** [comment]
Why is username the key field? Isn't there an ID for the request that is independent of username, that we can use to distinguish the requests? That would allow us to have 1 request for that username that is deleted/denied, while another is pending.

I agree however that there should never be 2 pending requests for the same username.

_Trac comment by ahelsing on 09-18-2013 at 11:18_

**Comment by ahelsing** [comment]
There is a request id but the uid is what links it to the ldap account, so that's what I was using as the key to both databases.  I could use the request id to differentiate between two entries with the same username, although now that I realize that there can be multiple request entries with the same username, I don't think its a good idea. I think its better to reject requests for a username that is already in the account database. 

_Trac comment by phelinek on 09-18-2013 at 11:29_

**Comment by ahelsing** [comment]
I guess I'm wondering about a request that we deny because it uses, say, a gmail email address. When the user comes back with a better email address, do we refuse to let them have their preferred username, because they tried with it before? That seems unfortunate.

_Trac comment by ahelsing on 09-18-2013 at 11:34_

**Comment by ahelsing** [comment]
OK that makes sense.  So then would we change the existing request to match the new request (new email address, etc) and change it from "Denied" to "Approved"?  Is that a problem if a different person requests a username that was denied for someone else? Or would we leave the old request in and have a second request with the same username? In which case I will go through the code and select from request ids instead of usernames.  

_Trac comment by phelinek on 09-18-2013 at 11:44_

**Comment by ahelsing** [comment]
I was thinking that this would be a 2nd request - as you say, preserving history

_Trac comment by ahelsing on 09-18-2013 at 11:52_

**Comment by ahelsing** [comment]
The only other problem is that if you delete the account, LDAP won't know which request to change.  But I can add a requestid column to the ldap table, fetch the id before deleting the account and then use the id for changing the status on the correct request.  So the solution would be:

1 - Change account request selects to use requestid.  
2 - Check for pending (REQUESTED,EMAILED LEADS,EMAILED REQUESTER) accounts when a new request is received and reject the request.  Maybe check for pending accounts with the same email while I'm at it.
3 - Create a duplicate request if the existing request is DENIED or DELETED (already does this).
4 - Add a column to LDAP for requestid, and use the requestid to change the status of existing account requests. 

_Trac comment by phelinek on 09-18-2013 at 12:55_

**Comment by ahelsing** [comment]
One more thing - existing accounts which are migrated to the new system will not have corresponding account requests, so the requestid will be empty for these. 

_Trac comment by phelinek on 09-18-2013 at 14:39_

**Comment by ahelsing** [comment]
In addition to the changes described above, I've added a new page to "fix" existing or historic accounts so they will work with the new system.  The page is:

manage/fix_accounts.html

If you push the initialize button, the script goes through all the existing ldap entries and:

1 - creates a dummy request if one doesn't exist. Title and Reason are set to HISTORIC.
2 - adds the new fields which include the request id to the ldap account entry

_Trac comment by phelinek on 09-20-2013 at 17:22_

**Comment by ahelsing** [comment]
Just chiming in belatedly: i think i agree with the consensus here and the solution which Pam implemented, and will check it out once i dig out of e-mail today.

Also: as far as i know, there shouldn't be any preexisting accounts with no history information, because geni-convert-logs takes care of that.  But i'll definitely try some actions on:
- Preexisting accounts which are active as of convert time
- Preexisting accounts which have been deleted as of convert time
  to make sure.

_Trac comment by chaos on 09-23-2013 at 10:08_

**Comment by ahelsing** [comment]
The historical data which is preserved in logs won't synchronize the account request database and the accounts database. That's where we need the fix_accounts function. 

_Trac comment by phelinek on 09-23-2013 at 13:31_

**Comment by ahelsing** [comment]
Oh, right, that script only affects the actions log.  Got it; thanks for spelling this out for me.

_Trac comment by chaos on 09-23-2013 at 13:36_

**Comment by ahelsing** [comment]
I tested the fix_accounts script, and it looks good.  There are currently 22 users active in aslund LDAP, and after i browse to fix_accounts, i see 22 entries in idp_account_request.

One thing i want to note is that if there are users who had previously had accounts and those accounts were deleted before migration, _those_ users will have entries in idp_account_actions but not in idp_account_request.  I tried making a new request with one such username, and it appears that the historical logs from the previous user show up for the new user, but nothing bad seems to happen at a glance.  So as far as i know this is harmless, but i wanted to mention it explicitly in case anyone thinks of a reason it would be a problem.

_Trac comment by chaos on 09-26-2013 at 16:21_

**Comment by ahelsing** [comment]
Hmm, as far as the initial issue for this ticket, though, i think it's still outstanding --- i redid the test in the description, and i still see two accounts with the same username in "Account Requests Waiting for Requester Response" on [https://shib-idp1.gpolab.bbn.com/manage/display_requests.php].

In the database:

```
accreq=> select * from idp_account_request where username_requested='chaos5';
 id | first_name |  last_name  |         email         | username_requested |    phone     |             password_hash              |   organization   |   title   | url |      reason       |         request_ts         | username_assigned |         created_ts         |   request_state   
----+------------+-------------+-----------------------+--------------------+--------------+----------------------------------------+------------------+-----------+-----+-------------------+----------------------------+-------------------+----------------------------+-------------------
 24 | Chaos      | Test User 5 | chaos5@gpolab.bbn.com | chaos5             | 617-873-2287 | {SSHA}FZi8vE8zGzIkgSkK7YYHyVXVWu/ommw3 | BBN Technologies | Test User |     | testing ticket 40 | 2013-09-26 20:41:41.792064 |                   | 2013-09-26 20:41:54.277193 | EMAILED_REQUESTER
 25 | Chaos      | Test User 5 | chaos5@gpolab.bbn.com | chaos5             | 617-873-2287 | {SSHA}xZqpHy+gZ2BvtdfwdCqJfRS59tHcLTbS | BBN Technologies | Test User |     | testing ticket 40 | 2013-09-26 20:42:34.681525 |                   |                            | EMAILED_REQUESTER
(2 rows)
```

_Trac comment by chaos on 09-26-2013 at 16:45_

**Comment by ahelsing** [comment]
Oh, i'm using version 0.5:

```
aslund,[~],16:45(0)$ dpkg -l geni-ar
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Cfg-files/Unpacked/Failed-cfg/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version        Description
+++-==============-==============-============================================
ii  geni-ar        0.5            GENI account request management
```

_Trac comment by chaos on 09-26-2013 at 16:45_

**Comment by ahelsing** [comment]
I'm having trouble reproducing the problem - when I tried your test the two accounts have different states as expected:

accreq=> select \* from idp_account_request where username_requested='dlastnam';

 id  | first_name | last_name |           email           | username_requested |    phone     |             password_hash              | organization | title | url | reason |         request_ts         | username_assigned |         created_ts         |   request_state  
-----+------------+-----------+---------------------------+--------------------+--------------+----------------------------------------+--------------+-------+-----+--------+----------------------------+-------------------+----------------------------+-------------------
 133 | Duptest    | Lastname  | phelinek15@gpolab.bbn.com | dlastnam           | xxx-xxx-xxxx | {SSHA}Ja0nApLI8zIHKt5qKLiQWEWSYO+2PgX8 | Org          | title |     | test   | 2013-09-27 02:37:12.718764 |                   | 2013-09-27 02:37:36.953513 | DELETED
 134 | Duptest    | Lastname  | phelinek15@gpolab.bbn.com | dlastnam           | xxx-xxx-xxxx | {SSHA}yLJniiExpLe5OQI2CJ0E0Cs3t5OYYEo/ | Org          | title |     | test   | 2013-09-27 02:39:56.882511 |                   |                            | EMAILED_REQUESTER
(2 rows)

_Trac comment by phelinek on 09-27-2013 at 08:47_

**Comment by ahelsing** [comment]
Hmm, i can reproduce it consistently.  Want to shoulder-surf my instance?

Is there any chance that this was fixed more recently than the geni-ar 0.5 package was created?  Tom's changelog, `geni-ar_0.5_i386.changes`, does contain:

```
   * Handle duplicate requests correctly (don't allow duplicate username or
     email if request is pending), associate request id with ldap account
     (#40)
```

and the other bugs named in the changelog were fixed in the version i have (i verified all those yesterday), but is there any chance something went awry in packaging?

Let me know how i can help figure out the discrepancy.

_Trac comment by chaos on 09-27-2013 at 09:54_

**Comment by ahelsing** [comment]
This was all in the same merge as the fix_accounts script, so it should be the right version.  I guess best is for me to stop by and take a look. 

_Trac comment by phelinek on 09-27-2013 at 10:18_

**Comment by ahelsing** [comment]
We just confirmed that the code I pushed to the release is missing the "where id='xxx' clause, although the clause is in the code I have running on ship-idp2.  I must have made the change in the wrong place and lost it in the merge for the package?  That explains why the behavior is different on my system and Chaos' system (and is easy to fix). 

_Trac comment by phelinek on 09-27-2013 at 10:39_

**Comment by ahelsing** [comment]
Verified; this is fixed in 0.6.

_Trac comment by chaos on 09-30-2013 at 15:00_
